### PR TITLE
Unstable merge: asset library with transparent compression, plus a stream of small changes

### PR DESCRIPTION
**Pull request by rasky**
This PR imports the asset library from the unstable branch. This is the subject of the first half of commits.

The asset library allows to load binary files from any filesystem with transparent compression, and is the basis of all data files management in the unstable branch, which is in fact compressed.

Compression works by wrapping a data file (think gzip), so `asset_load()` will transparently decompress it while loading it into RDRAM. Moreover, `asset_fopen()` uses some libc magic to create a `FILE*` that transparently decompresses while reading, to allow compressing data files that need to be parsed at runtime.

There are two compression levels at the moment: 

 * Level 1 is based on LZ4, which features incredible fast decompression speed with decent compression ratios. It is so fast that it's actually faster to load and decompress a compressed file, rather than loading it uncompressed (!). On the unstable branch, we compress all assets by default with level 1 because of this.
 * Level 2 is based on LZH5, part of the LHA family. It offers better-than-gzip ratios on small files, with a slower decompression. The decompressor is based on open source code that has been optimized for N64, but it's still not as fast as it could. Specifically, the Huffman decompression is still made bitwise rather than bytewise and could use some speed up.

To compress a binary file, the user can use the new tool `mkasset`. In the future, asset-generating tools (like the new version of mksprite) will transparently offer compression capabilities.

See this table for more data (c0 = uncompress, c1 = lz4, c2 = lzh5)
```
File                        : KiB  | Ratio   | Load+Dec (ms)
--------------------------------------------------------------
fc_200-fs8.sprite.c0        :   31 |  100.0% |     11.4 
fc_200-fs8.sprite.c1        :   31 |   32.9% |      5.9 
fc_200-fs8.sprite.c2        :   31 |   25.7% |     19.6 
font.sprite.c0              :   22 |  100.0% |      7.8 
font.sprite.c1              :   22 |   19.8% |      4.1 
font.sprite.c2              :   22 |   14.4% |      9.4 
hoi-fs8.sprite.c0           :  125 |  100.0% |     41.1 
hoi-fs8.sprite.c1           :  125 |   11.5% |     14.3 
hoi-fs8.sprite.c2           :  125 |    8.9% |     29.3 
lens_64-fs8.sprite.c0       :   16 |  100.0% |      7.1 
lens_64-fs8.sprite.c1       :   16 |   26.7% |      4.4 
lens_64-fs8.sprite.c2       :   16 |   23.8% |      9.4 
lenspic_256-fs8.sprite.c0   :   64 |  100.0% |     22.0 
lenspic_256-fs8.sprite.c1   :   64 |   50.5% |     10.5 
lenspic_256-fs8.sprite.c2   :   64 |   44.2% |     58.2 
prax4.fli.c0                :  327 |  100.0% |    104.5 
prax4.fli.c1                :  327 |   63.2% |     85.7 
prax4.fli.c2                :  327 |   51.2% |    316.9 
room_normal.ci8.sprite.c0   :  125 |  100.0% |     41.6 
room_normal.ci8.sprite.c1   :  125 |   31.7% |     19.2 
room_normal.ci8.sprite.c2   :  125 |   26.5% |     67.2 
srtitle400-fs8.sprite.c0    :  125 |  100.0% |     40.9 
srtitle400-fs8.sprite.c1    :  125 |   28.6% |     18.4 
srtitle400-fs8.sprite.c2    :  125 |   22.9% |     60.8 
troll-fs8.sprite.c0         :  125 |  100.0% |     41.9 
troll-fs8.sprite.c1         :  125 |   32.6% |     25.1 
troll-fs8.sprite.c2         :  125 |   24.7% |     61.5 
```

The rest of the PR is a stream of small semi-obvious commits that just minimizes the differences with unstable.

### CODE REVIEW

File: include/asset.h [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+ * by #asset_fopen will assert on seek, even if the file is not compressed
+ * (so that the user code will be ready for adding compression at any time).
+ * 
+ * If you know that the file will never be compressed and you absolutely need
+ * to freely seek, simply use the standard fopen() function.
+ * 
+ * ## Asset compression
+ * 
+ * All tools in libdragon that produce assets already integrate compression
+ * support compatible with the asset libraries. To compress your own data
```
Comment (snacchus) [inline include/asset.h]: Is this actually true on trunk right now?

File: include/asset.h [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+/**
+ * @brief Load an asset file (possibly uncompressing it)
+ * 
+ * This function loads a file from a file system (eg: from ROM or SD).
+ * If the file was compressed using the mkasset tool, it will be
+ * automatically uncompressed.
+ * 
+ * @param fn        Filename to load (including filesystem prefix)
+ * @param sz        Pointer to an integer where the size of the file will be stored
```
Comment (snacchus) [inline include/asset.h]: `sz` can be `NULL`, right? Should probably mention this here.

File: src/compress/ringbuf.c [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+void __ringbuf_copy(decompress_ringbuf_t *ringbuf, int copy_offset, uint8_t *dst, int count)
+{
+    int ringbuf_copy_pos = (ringbuf->ringbuf_pos - copy_offset) & (RING_BUFFER_SIZE-1);
+    int dst_pos = 0;
+    while (count > 0) {
+		int wn = count;
+        wn = wn < RING_BUFFER_SIZE - ringbuf_copy_pos     ? wn : RING_BUFFER_SIZE - ringbuf_copy_pos;
+		wn = wn < RING_BUFFER_SIZE - ringbuf->ringbuf_pos ? wn : RING_BUFFER_SIZE - ringbuf->ringbuf_pos;
```
Comment (snacchus) [inline src/compress/ringbuf.c]: Some whitespace issues here

File: include/asset.h (hunk 1 above)
Comment (rasky) [reply include/asset.h]: removed that sentence

File: include/asset.h (hunk 2 above)
Comment (rasky) [reply include/asset.h]: Done

File: src/compress/ringbuf.c (hunk 3 above)
Comment (rasky) [reply src/compress/ringbuf.c]: Fixed

**Comment by rasky** [comment]
Added also the new API `asset_init_compression()` as discussed on Discord, to avoid linking all compression levels unless really necessary.

**Pull request merged**